### ISSUE TITLE
Unsatcore with Java crashes

### ISSUE THREAD

**Issue by GerhardSchellhorn** [CVC4/CVC4#2989]
I'm trying to get cvc4-1.6 to work (under Scala) with the Java bindings.
My configuration is Java-8-openjdk, ubuntu18.04
However I can't find any documentation on how to properly name formulas, or how
to get the names of used formulas in the unsatcore. There seems to be an iterator,
but calling it produces crashes, see the small program below (which proves "A or B"
from axiom "A and B")
So my questions are: Am I doing something wrong to get these crashes?
What is the correct way  to get the names of axioms / goal in the unsatcore?

```
import java.io.*;
import java.util.Iterator;
import edu.nyu.acsys.CVC4.*;

public class CVC4test {


    // adjust this path
  public static String libpath = "/home/cvc4/lib/jni";
  public static void main(String[] args) {

   /** Load CVC4's native library */
   File libraryPath = new java.io.File(libpath, System.mapLibraryName("cvc4jni"));
   if (!libraryPath.exists()) {
      System.err.println("cvc4 library not found");
      System.exit(-1);
   }
   String libraryPathName = libraryPath.getAbsoluteFile().getPath();
   System.out.println("smt: loading " + libraryPathName);
   System.load(libraryPathName);
   System.out.println("smt: loaded " + libraryPathName);

   ExprManager em = new ExprManager();
   SmtEngine smt = new SmtEngine(em);

   smt.setOption("incremental", new SExpr(true));
   smt.setOption("stats", new SExpr(true));
   smt.setOption("produce-unsat-cores", new SExpr(true));
   smt.setOption("tlimit", new SExpr(10000));
   smt.setOption("e-matching",new SExpr(true));
   smt.setOption("macros-quant", new SExpr(true));
   smt.setLogic("ALL_SUPPORTED");

   Type bool_sort = em.booleanType();
   Expr Avar = em.mkVar("A", bool_sort);
   Expr Bvar = em.mkVar("B", bool_sort);

   Expr AandB = em.mkExpr(Kind.AND, Avar, Bvar);
   Expr AorB = em.mkExpr(Kind.OR, Avar, Bvar);
   Expr goal = em.mkExpr(Kind.NOT, AorB);
   // is this the way to name a formula?
   //smt.setExpressionName(AandB,"axiom");
   //smt.setExpressionName(goal,"goal");
   smt.assertFormula(AandB);

   smt.assertFormula(goal);
   Result res = smt.checkSat(); // result is unsat
   System.out.println("CVC4 reports: " + res);
   Result.Sat result = res.isSat();
   if (result.toString() == "UNSAT") {
       System.out.println("Result is UNSAT");
   }
   else System.out.println("Result is not UNSAT");

   UnsatCore unsatCore = smt.getUnsatCore();
   System.out.println("CVC4 unsatcore: " + unsatCore);
   // this code crashes
   Iterator<Expr> it = unsatCore.iterator();
   while (it.hasNext()) {
       System.out.println("Corefma:" + it.next());
     }
  }

}
```

**Comment by 4tXJ7f** [comment]
Thank you very much for your report @GerhardSchellhorn. Unfortunately, the documentation on the unsat cores API is really lacking. I was able to reproduce the crash on current master and this is definitely a bug. I also found that `System.out.println("CVC4 unsatcore: " + unsatCore);` did not really print the unsat core, just the class name. I found a fix for the latter issue but am still working on the first one.

If you want to use named expressions, you actually have to set the language to SMT-LIB 2 by doing `smt.setOption("lang", new SExpr("smt2"))`. The reason for this is that the default language is the CVC language which has no concept of named expressions.

Side-note: CVC4 1.6 is not the most recent stable release. We recently released CVC4 1.7.

I'll keep you posted.

**Comment by GerhardSchellhorn** [comment]
ok, thanks for the response. I'll switch to 1.7 then. After setting the lang-option, it would still help me then if you let me know how to give names to axioms  (as in the code I commented out?).

**Comment by 4tXJ7f** [comment]
I have fixed the issue in PR #3000. If you need the fix now, you could try using the branch. I am not quite sure yet when we will be able to merge the PR because we currently have a code freeze due to the deadline for SMT-COMP coming up.

**Comment by 4tXJ7f** [comment]
Alright, the fix is now in master and an example for interacting with the unsat cores can be found [here](https://github.com/CVC4/CVC4/blob/62d361071ea54c9b7cba882313ab4dedef6f1286/examples/api/java/UnsatCores.java).
